File: benchmarks/polyglot/bench.zig
const std = @import("std");

// Zig 0.15 + macOS 26 linker workaround: use extern write(2) for stdout.
// Compile: zig build-obj bench.zig -OReleaseFast && cc bench.o -o bench_zig -lSystem
extern "c" fn write(fd: c_int, buf: [*]const u8, count: usize) isize;

fn print(comptime fmt: []const u8, args: anytype) void {
    var buf: [512]u8 = undefined;
    const result = std.fmt.bufPrint(&buf, fmt, args) catch return;
    _ = write(1, result.ptr, result.len);
}

fn fib(n: i32) i32 {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

fn benchFibonacci() void {
    const start = std.time.milliTimestamp();
    const result = fib(40);
    const elapsed = std.time.milliTimestamp() - start;
    print("fibonacci:{d}\n", .{elapsed});
    print("  checksum: {d}\n", .{result});
}

fn benchLoopOverhead() void {
    const start = std.time.milliTimestamp();
    var sum: f64 = 0.0;
    var i: i32 = 0;
    while (i < 100_000_000) : (i += 1) {
        sum += 1.0;
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("loop_overhead:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{sum});
}

fn benchArrayWrite() void {
    const allocator = std.heap.page_allocator;
    const arr = allocator.alloc(f64, 10_000_000) catch return;
    defer allocator.free(arr);

    const start = std.time.milliTimestamp();
    for (0..10_000_000) |i| {
        arr[i] = @as(f64, @floatFromInt(i));
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("array_write:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{arr[9_999_999]});
}

fn benchArrayRead() void {
    const allocator = std.heap.page_allocator;
    const arr = allocator.alloc(f64, 10_000_000) catch return;
    defer allocator.free(arr);

    for (0..10_000_000) |i| {
        arr[i] = @as(f64, @floatFromInt(i));
    }

    const start = std.time.milliTimestamp();
    var sum: f64 = 0.0;
    for (0..10_000_000) |i| {
        sum += arr[i];
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("array_read:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{sum});
}

fn benchMathIntensive() void {
    const start = std.time.milliTimestamp();
    var result: f64 = 0.0;
    var i: i32 = 1;
    while (i <= 50_000_000) : (i += 1) {
        result += 1.0 / @as(f64, @floatFromInt(i));
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("math_intensive:{d}\n", .{elapsed});
    print("  checksum: {d:.6}\n", .{result});
}

const Point = struct {
    x: f64,
    y: f64,
};

fn benchObjectCreate() void {
    const start = std.time.milliTimestamp();
    var sum: f64 = 0.0;
    var i: i32 = 0;
    while (i < 1_000_000) : (i += 1) {
        const fi = @as(f64, @floatFromInt(i));
        const p = Point{ .x = fi, .y = fi * 2.0 };
        sum += p.x + p.y;
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("object_create:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{sum});
}

fn benchNestedLoops() void {
    const n: usize = 3000;
    const allocator = std.heap.page_allocator;
    const arr = allocator.alloc(f64, n * n) catch return;
    defer allocator.free(arr);

    for (0..n * n) |i| {
        arr[i] = @as(f64, @floatFromInt(i));
    }

    const start = std.time.milliTimestamp();
    var sum: f64 = 0.0;
    for (0..n) |i| {
        for (0..n) |j| {
            sum += arr[i * n + j];
        }
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("nested_loops:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{sum});
}

fn benchAccumulate() void {
    const start = std.time.milliTimestamp();
    var sum: f64 = 0.0;
    var i: i64 = 0;
    while (i < 100_000_000) : (i += 1) {
        sum += @as(f64, @floatFromInt(@mod(i, 1000)));
    }
    const elapsed = std.time.milliTimestamp() - start;
    print("accumulate:{d}\n", .{elapsed});
    print("  checksum: {d:.0}\n", .{sum});
}

pub fn main() void {
    benchFibonacci();
    benchLoopOverhead();
    benchArrayWrite();
    benchArrayRead();
    benchMathIntensive();
    benchObjectCreate();
    benchNestedLoops();
    benchAccumulate();
}

File: benchmarks/polyglot/bench.java
public class bench {

    static int fib(int n) {
        if (n < 2) return n;
        return fib(n - 1) + fib(n - 2);
    }

    static void benchFibonacci() {
        long start = System.currentTimeMillis();
        int result = fib(40);
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("fibonacci:" + elapsed);
        System.out.println("  checksum: " + result);
    }

    static void benchLoopOverhead() {
        long start = System.currentTimeMillis();
        double sum = 0.0;
        for (int i = 0; i < 100_000_000; i++) {
            sum += 1.0;
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("loop_overhead:" + elapsed);
        System.out.printf("  checksum: %.0f%n", sum);
    }

    static void benchArrayWrite() {
        double[] arr = new double[10_000_000];
        long start = System.currentTimeMillis();
        for (int i = 0; i < 10_000_000; i++) {
            arr[i] = (double) i;
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("array_write:" + elapsed);
        System.out.printf("  checksum: %.0f%n", arr[9_999_999]);
    }

    static void benchArrayRead() {
        double[] arr = new double[10_000_000];
        for (int i = 0; i < 10_000_000; i++) {
            arr[i] = (double) i;
        }
        long start = System.currentTimeMillis();
        double sum = 0.0;
        for (int i = 0; i < 10_000_000; i++) {
            sum += arr[i];
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("array_read:" + elapsed);
        System.out.printf("  checksum: %.0f%n", sum);
    }

    static void benchMathIntensive() {
        long start = System.currentTimeMillis();
        double result = 0.0;
        for (int i = 1; i <= 50_000_000; i++) {
            result += 1.0 / (double) i;
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("math_intensive:" + elapsed);
        System.out.printf("  checksum: %.6f%n", result);
    }

    static class Point {
        double x;
        double y;

        Point(double x, double y) {
            this.x = x;
            this.y = y;
        }
    }

    static void benchObjectCreate() {
        long start = System.currentTimeMillis();
        double sum = 0.0;
        for (int i = 0; i < 1_000_000; i++) {
            Point p = new Point((double) i, (double) i * 2.0);
            sum += p.x + p.y;
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("object_create:" + elapsed);
        System.out.printf("  checksum: %.0f%n", sum);
    }

    static void benchNestedLoops() {
        int n = 3000;
        double[] arr = new double[n * n];
        for (int i = 0; i < n * n; i++) {
            arr[i] = (double) i;
        }
        long start = System.currentTimeMillis();
        double sum = 0.0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sum += arr[i * n + j];
            }
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("nested_loops:" + elapsed);
        System.out.printf("  checksum: %.0f%n", sum);
    }

    static void benchAccumulate() {
        long start = System.currentTimeMillis();
        double sum = 0.0;
        for (int i = 0; i < 100_000_000; i++) {
            sum += (double) (i % 1000);
        }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("accumulate:" + elapsed);
        System.out.printf("  checksum: %.0f%n", sum);
    }

    public static void main(String[] args) {
        benchFibonacci();
        benchLoopOverhead();
        benchArrayWrite();
        benchArrayRead();
        benchMathIntensive();
        benchObjectCreate();
        benchNestedLoops();
        benchAccumulate();
    }
}
